models/wiki_card_db.py

```diff
@@ -322,85 +324,289 @@ def clean_fields(cards, locales):
+
+
+def process_aa_card_data(card_data):
+    """This card data has gone through bifurcation, so we expect it to
+    represent an AA card page. Here we can apply simple single-card
+    processing, like for X power to integer 0 to prevent schema type failures
+    on the site."""
+
+    # Skip null power like for actions.
+    if card_data["power"]:
+        # Motivating case is "X" power creatures, but AA schema
+        # needs an integer value.
+        try:
+            card_data["power"] = int(card_data["power"])
+        except ValueError:
+            card_data["power"] = 0
+
+
+def bifurcate_data(card_datas):
+    """For a set of cards, isolate them into the list of cards
+    we want to put onto Archon Arcana as individual card pages. Some cards with the same name in the db
+    will be merged into one card (such as a card that can appear in multiple houses), while others will be
+    separated into multiple cards (Anomaly version of a card and the non-Anomaly version of the card)
+    Returns a list of card data dictionaries that something else can process
+
+    Input: card_datas is a list of card dictionaries
+    Each dict is one MV card JSON all with the same card title.
+
+    Output: a list of card dict where we've done any of:
+      - merged dicts, like two halves of gigantic into one dict
+      - split the list into distinct dicts, like an anomly or
+        redemption variant distinct from the normal card"""
+    # TODO clean up when we refactor card db stuff to a class
+    # TODO - note, to combine houses, this method requires the batch to include the complete set of cards of a given name
+
+    if not card_datas:
+        return []
+
+    if card_datas[0]["card_title"] == "Dark Æmber Vault":
+        import pdb; pdb.set_trace()
+
+    if len(card_datas) == 1:
+        return card_datas
+
+    (
+        is_giant, merged_halves
+    ) = bifurcate_giants(card_datas)
+    if is_giant:
+        return merged_halves
+
+    (
+        has_anomaly, anomalies, other
+    ) = bifurcate_anomalies(card_datas)
+    if has_anomaly:
+        return anomalies + bifurcate_data(other)
+
+    (
+        has_redemption, redemption, other
+    ) = bifurcate_redemption(card_datas)
+    if has_redemption:
+        return redemption + bifurcate_data(other)
+
+    # Do special multi-house cases before this, like Redemption.
+    (
+        multi_house, merged
+    ) = bifurcate_multi_house(card_datas)
+    if multi_house:
+        return merged
+
+    # We shouldn't fall to this case and still have
+    # multiple types.
+    types = set([card["card_type"] for card in card_datas])
+    if len(types) > 1:
+        card_title = card_datas[0]["card_title"]
+        print(card_datas)
+        raise Exception(f"Unknown type mismatch {card_title} {types}")
+
+    # Otherwise the batch doesn't need bifurcation.
+    # TODO: what does it make here? Same card name from
+    # different sets? We build set data differently elsewhere?
```

Comment:
Info:

The input and output of this function is essentially the same: a list of related card data dictionaries that are intended to map to a SINGLE AA page. The sole purpose is to either return the list without processing, so that all of the data will be made into a single page; OR, split up the dictionaries to make multiple AA pages, sending the appropriate subset to each page.

So:
   [Curse of Forgetfulness (WoE data), Curse of Forgetfulness (GR data), Curse of Forgetfulness (ToC data)]
becomes:
   [Curse of Forgetfulness (Anomaly) (WoE data)], 
   [Curse of Forgetfulness (GR data), Curse of Forgetfulness (ToC data)]
   
But, [Chuff Ape (CotA data), Chuff Ape (VM2023 data)] remains unchanged. HOW to merge the multiple card data into a single page happens elsewhere - after my_card_db.json is generated. 

In the json, each card is a dictionary of all of the set numbers it has data for. The sets are combined in get_cargo: https://github.com/saluk/archonarcanabots/blob/241156132a9ba837e32e2c75fbb22ed380928786/models/wiki_card_db.py#L701 (though that function takes an argument called "card" it's really "card data from the my_card_db.json file with all of its set keys)

Reply:
Understood! This is where I needed to take time, do the refactor, better understand for now and future sets.